### PR TITLE
Add Czech rules --full-wiki-extraction=cs

### PR DESCRIPTION
**Pull request by comodoro**
Czech rules for discussion. The characters might seem too restricted, but still it yields more than 260k sentences

- How many sentences did you get at the end?

Cca. 260000

- How did you create the blacklist file?

The way mentioned in the readme, filtering out words with frequency 100 and less

- Get at least 3 different native speakers (ideally linguistics) to review a random sample of 100-500 sentences and estimate the average error ratio and comment (or link their comment) in the PR.

I have posted a sample to https://gist.github.com/comodoro/7fc7cd64b5b7df6a3b033dca1146f8ff. The error ratio seems quite low, certainly lower than in the sentence collector, I do not know any linguists, but will try to talk to someone. Any Czech speakers who see this are welcome to comment:)

### CODE REVIEW

File: src/rules/czech.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+allowed_symbols_regex="[A-Za-zěščřžýáíéóďťňĚŠČŘŽÝÁÍÉÓĎŤŇäöüÄÖÜ‚–\\. ]"
```
Comment (MichaelKohler) [inline src/rules/czech.toml]: would it make sense to allow quotation symbols? Would that change anything in terms of resulting total sentences?

File: src/rules/czech.toml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+min_trimmed_length = 6
+min_word_count = 1
+max_word_count = 14
+min_characters = 0
+may_end_with_colon = false
+quote_start_with_letter = true
+needs_punctuation_end = false
+needs_letter_start = true
+needs_uppercase_start = true
+even_symbols = ["\""]
```
Comment (MichaelKohler) [inline src/rules/czech.toml]: minor: given that the `allowed_symbols_regex` rule does not allow these quotes, this check will never be run, and could therefore be removed.

**Comment by Adrijaned** [comment]
Went through the first 300 sentences in the gist posted, noticed these erronenous sentences:
`Celková hodnota obsahu cache se nesmí výměnou snížit.-` "cache" is not really a czech word
`Vznikaly také první umělecké školy a další umělecké instituce jako například výstavy a festivaly.`- missing comma before "jako"
`O dějinách historiografie` - not a sentence, clearly a heading
`Tíhové zrychlení tedy obecně závisí na` - missing half of a sentence
`Slovo se používá i v přeneseném významu pro` - same as above
`Mj. namaloval série pohlednic s texty v esperantu nebo na motivy esperanta.` - contains abbreviation Mj. - mimo jiné
`Mezi nejznámější patří` - missing most of the sentence
`Podle závislosti na čase lze elektrické pole rozdělit na` - same as above
`Adolfa Hitlera.` - missing first part of the sentence

That makes up for an error rate of roughly 3%? It could be probably brought down a bit by requiring the sentences to end in punctuation, as all sentences I found cut unnaturally were missing it, and there were no sentences without the punctuation that I would consider correct.

File: src/rules/czech.toml (hunk 1 above)
Comment (Adrijaned) [inline src/rules/czech.toml]: You missed `úůÚŮ`, and I'd perhaps remove `ÄÖÜäöü` - they are only used in foreign names and some might have issues pronouncing them, but don't consider me an authority on this.

Comment (comodoro) [reply src/rules/czech.toml]: A run with just adding double quotes resulted in 5k more sentences, I will check on their error rate

Comment (comodoro) [reply src/rules/czech.toml]: Thanks, these are all good suggestions, plus the above, I will incorporate them when I have time to make a next run.

**Comment by MikkCZ** [comment]
There are some weird sentences like `Adolfa Hitlera.` or `Žije zde obyvatel.`, which nobody would say as they are, but for the purpose of voice data set it seems OK too me. They are Czech.

Most of what I noticed are incomplete sentences, which would easilly (I assume) be filtered by dropping everything not ending with `.`, `?` or `!`.

**Comment by comodoro** [comment]
I have updated the rules and the gist with the above suggestions. Currently about 329k sentences are generated.

File: src/rules/czech.toml (hunk 1 above)
Comment (comodoro) [reply src/rules/czech.toml]: Added in the last commit

File: src/rules/czech.toml (hunk 2 above)
Comment (comodoro) [reply src/rules/czech.toml]: OK, I am keeping them

File: src/rules/czech.toml (hunk 1 above)
Comment (comodoro) [reply src/rules/czech.toml]: Added except for Ů that would anyways always get removed as a non-word-leading capital

**Comment by comodoro** [comment]
Sorry for following up late, I have tried to contact someone with linguistic background, without much success, but the main problem for ASR is still imho foreign names and words, which I do not see how to deal with. These are still higher quality than what came through sentence collector, I would say they should be used, particularly as I have encountered duplicates already. What else can I do? @MichaelKohler @Adrijaned

**Comment by MichaelKohler** [comment]
Can I also ask you to do the following, to make sure you can profit from the [automatic sample extraction](https://discourse.mozilla.org/t/scraper-automatic-sample-sentences-extracted-in-pull-request/55217) we just introduced?

* Update your branch with the latest code from the `master` branch
* Rename `src/rules/czech.toml` to `src/rules/cs.toml`
* Rename `src/rules/disallowed_words/czech.toml` to `src/rules/disallowed_words/cs.toml`

Also note that the local command for extraction will now be:

```
cargo run -- extract -l cs -d path/to/files
```

Happy to answer any question you may have and thanks for your efforts!

Also, did you have any luck finding reviewers?

**Comment by comodoro** [comment]
I have updated the rules. Allowing more paired symbols with the new matching symbol check, which is not really worth it, currently about 333000 sentences. No linguists found, I will try some general forums at least. New sample is at https://gist.github.com/comodoro/a2606297fb77c2fa7dcc399ed6559c70

**Comment by MikkCZ** [comment]
Despite being native Czech speaker, I don't consider myself as linguist. However I try to reach more people around @l10ncz and hopefully someone will feel expert enough to have a look and comment here.

@MichaelKohler what are the acceptance criteria to meet?

**Comment by MichaelKohler** [comment]
If you can't find an expert, I think localization contributors are a good source indeed. The requirement is to have a maximum error rate of 5-7%. Anything lower is great, but very hard to achieve.

**Comment by comodoro** [comment]
Actually localization is one of my occupations:) FWIW I have quickly skimmed the last 100 sentences and there was one sentence with English words in it and several with foreign names, so I would guess about 1-5%, depending on the attitude to foreign names.

**Comment by MichaelKohler** [comment]
The following template might help to come up with an exact error rate: https://docs.google.com/spreadsheets/d/1l_5bP01ggRbVcwBosBcTCT0zkB6cABPTsnQzKAJdlJU/edit#gid=0

**Comment by comodoro** [comment]
If you go only for uncommon/unpronounceable foreign names, I don't think they are in the last 100, meaning ~1%, but some are bound to be in the rest of the text,

**Comment by Adrijaned** [comment]
Reviewing sentences 500-700 from the last gist posted, issues I noticed:
`Je také důležitým jazykem v Louisianě.` - some may have issues pronouncing Louisiana
`Aktualizace umožňuje (kromě oprav chyb) například použití novějšího modelu procesoru.` - not sure about the parentheses
`Data General.` - english name of software, probably
`Ultrazvukové vlny procházejí tělem a odrážejí se od orgánů (akustická impedance).`
`Stavové veličiny se dělí na vnější (extenzivní) a vnitřní (intenzivní).`
`Poručík a Adelheid spolu stráví noc v posteli (předtím ji tam posílá).`
`Natáčelo se i v samotném objektu Studia Barrandov (interiéry počítačového střediska).`
`Navzdory rozdílům mezi jednotlivými dialekty (nebo jazyky) si většina mluvčích různých dialektů navzájem rozumí.`
`Kostelník z města (Vlastimil Brodský) se stane falešným farářem ve vesnici.` - parentheses
`Krátce na to zavedl teoretický fyzik John Wheeler výraz „černá díra“.` - problematic name
`Na její místo nastoupí nový učitel Igor Hnízdo (Jan Tříska).` - parentheses

That's slightly above 5%, but it's been while since I worked on this, so I might be misremembering some rules. Also, I might have just picked some wrong part.

**Comment by comodoro** [comment]
Thanks. I was under the impression that parenthenses are not a problem, but I may be mistaken,Some languages have them removed as well. Perhaps @MichaelKohler could clarify?

**Comment by MichaelKohler** [comment]
> Thanks. I was under the impression that parenthenses are not a problem, but I may be mistaken,Some languages have them removed as well. Perhaps @MichaelKohler could clarify?

In both [Sentence Collector rules](https://github.com/Common-Voice/sentence-collector/blob/master/shared/validation/languages/en.js#L22) and [Sentence Extractor rules](https://github.com/Common-Voice/cv-sentence-extractor/blob/master/src/rules/en.toml#L11) we disallow parenthesis at least for English/default.

**Comment by comodoro** [comment]
All right, I have removed the parentheses. Posted new gist to https://gist.github.com/comodoro/eafda25b5e224666c3d3d1fb2e761884. About 330000 sentences overall. At a glance there are still ~1-2% foreign names and words.

**Comment by Adrijaned** [comment]
Issues found in posted gist, sentences 1-500:
```
Celková hodnota obsahu cache se nesmí výměnou snížit.
Může v cache ale zanechat jiný takový drobný předmět.
Mnoho steam her bylo přeneseno na Linux.
Temné období a počíná doba železná.    - my guess is this was cut after
'tzv.' ('so-called')
Jako Němci museli nastoupit do wehrmachtu a bojovat ve válce na straně
Hitlera.   -Wehrmacht might be problematic to pronounce for some
Mezi slezské malíře se často uvádí Adolph von Menzel.
Adolfa Hitlera.
Žije zde obyvatel.     - missing number
Vysloužila si přezdívku "Radar range".
Protože dosah gravitační síly je nekonečný i gravitační pole je vlastně
nekonečné.   - missing comma
Dnes používáme označení jednotku Hz a její násobky.
Vídeňském kroužku.    - my guess is this was cut after 'tzv.' ('so-called')
Podle se Maďarsko dělí ve třech stupních na statistické regiony.    -
missing name or other source
Závislost tlaku během proudění ideální kapaliny popisuje Bernoulliho
rovnice.
Závislost tlaku a rychlosti proudění vychází z Bernoulliho rovnice.
```

**Comment by comodoro** [comment]
Thanks, that would make the (conservative, I would say) error rate some 3.2%, if I count correctly.

Similarly for 501-600:

Je také důležitým jazykem v Louisianě.
Ukládání data a instrukcí.
Firmware je v současných elektronických zařízeních běžně používán.
V osobních počítačích je firmware často používán.
Za císaře Augusta dále proběhla menší změna v názvech délkách měsíců.

(Very conservative 5%)

**Comment by Adrijaned** [comment]
@comodoro of what you posted, only 1 and 5 would be an error in my review. I don't see any issue at all with 2 (it looks like it might be coming from some planning software), and there isn't really a way to mispronounce "firmware" (+"software" is a common word enough)

**Comment by comodoro** [comment]
I agree, it is very conservative. I doubt "Louisiana" is harder to pronounce than "firmware". Only the last one is really a bit problematic imho.

**Comment by dlukes** [comment]
Native speaker and linguist (@czcorpus) here :) I went through sentences 601--1000 from the last gist, details below.

For the purpose of estimating the error rate, I would only consider the first two headings as relevant, so that's **5 out of 400 = 1.25%**.

### Sentence extraction errors

These fragments are likely to be sentence extraction errors:

```
Novell.
Marketingové označení prakticky vyjadřuje max.
Geografické společnosti ve Vídni.
```

### Proofreading errors

These pertain to the quality of the data source, not the extraction process. I found two if I'm being very nitpicky:

```
# skupinou -> skupinu
Zdaleka největší skupinou tvoří Angličané.
# slightly weird word order
V prosinci svobodně zvolený magistrát byl nucen přesídlit do Západního Berlína.
```

### Loanwords, foreign names, code switching etc.

These might be hard to pronounce for some speakers, since they don't follow the general rule that in Czech, pronunciation can usually be easily inferred from spelling (unlike in English). On the other hand, their presence is a reality in modern Czech, so trying to systematically weed them out is not a good idea IMHO. Also, the people who are likely to volunteer to read these sentences can be reasonably expected to able to cope with them, I think. And obviously, many of these are fairly common and well-known.

So in summary, I wouldn't worry about these too much :) I just listed them because I noticed that the previous comments paid attention to them as well.

```
Tento termín byl poprvé použit psychologem a filozofem Williamem Jamesem.
Krátce na to zavedl teoretický fyzik John Wheeler výraz „černá díra“.
Dalším známým zástupcem je i border kolie.
Napsal hru nazvanou Space Travel.
Krátce na to zavedl teoretický fyzik John Wheeler výraz „černá díra“.
Dalším známým zástupcem je i border kolie.
Napsal hru nazvanou Space Travel.
Nejvýznamnějším dramatikem tohoto období byl Angličan William Shakespeare.
Nejdůležitějším místem vývoje muzikálu byla newyorská Broadway a její divadla.
Většina celosvětově proslulých muzikálů měla premiéru na Broadwayi.
Fragmenty této rockové opery hraje v koncertní podobě skupina Tábor Superstar Band.
Jazz rap spojil jazz a hip hop.
Později jazz přebíral i prvky populární hudby.
Je pohřben na hřbitově v Montmartru v Paříži.
Tento debut ji proslavil.
V Dijonu vystudovala divadelní konzervatoř.
V moderní adaptaci románu Alexandra Dumase vytvořila postavu Lindy.
Výjimkou je postava Alexandra Dumase.
Kvůli práci na vylepšování parního stroje se však Watt zadlužil.
Po návratu do Glasgow se stal univerzitním mechanikem.
Ale Andersen odchodem z Odense pohřbil celé své dětství.
Sám Jacob se nikdy neoženil.
Microsoft poskytuje studentské slevy na některé své produkty.
Společně vystoupili také na několika málo koncertech v New Yorku a Los Angeles.
Hudební inspiraci nabral hlavně v jazzu.
Ani v moderní literatuře není Poe zcela bez vlivu.
Poe se ve finanční krizi zapsal na vojenskou akademii ve West Pointu.
Po překročení Seiny pronikli Spojenci do Belgie a k německým hranicím.
Během vyučování si dělají tři studenti legraci ze smrti Reinharda Heydricha.
```

### Blacklist template sentences?

I found two instances of template sentences presumably extracted from a family of articles built upon the same template:

```
Svátek (má|slaví) {name or names}.
Jedná se o {numeral} století {numeral} tisíciletí.
```

Maybe it would make sense to blacklist these sentence templates, if it's possible to do so? Linguistically though, there's nothing wrong with the sentences, they're just repetitive, which may or may not be an issue.

### Odds and ends

People not familiar with astrophysics might think "Ia" is "eye ay", not "one ay":

```
Typ supernov Ia uvolňuje největší množství energie mezi všemi ostatními známými třídami supernov.
```

**Comment by comodoro** [comment]
Thank you, this is great work! Regarding blacklisting sentences, there is about 149 "Svátek má" overall (in 330000), of which significant part got into the sample. I have not found the other one, numerals are filtered.

@MichaelKohler Is this satisfactory? Can it be merged?

**Comment by dlukes** [comment]
> I have not found the other one, numerals are filtered.

Sorry for the confusion, I meant numeral _words_, i.e. written out in letters :) E.g. “Jedná se o třetí století prvního tisíciletí.” and the like.

**Comment by comodoro** [comment]
Oh, I see. This may not be a template, search for `Jedná se o .* století`` yielded 9 results, the last one on line 1004.

**Comment by MichaelKohler** [comment]
> @MichaelKohler Is this satisfactory? Can it be merged?

Can you add a final comment answering the following questions?

* How many sentences did you get at the end?
* How did you create the blacklist file?
* What's the final error rate?

Then @nukeador can approve this.

**Comment by comodoro** [comment]
> * How many sentences did you get at the end?
330338

> * How did you create the blacklist file?
Words with frequency 80 and less, the recommended method

> * What's the final error rate?
That very much depends on who estimates it. Based on the three reviews of the last iteration, something like (3.2 * 500 + 5 * 100 + 1.25 * 400) / 1000 = 2.6%.

**Comment by nukeador** [comment]
This looks good to go.

Thanks so much everyone involved in this work!

@MichaelKohler let's queue this for extraction and integration. If this is merged before Tuesday it can get live with the website deployment next week.

Cheers.

**Comment by MichaelKohler** [comment]
Let's see if the automatic extraction job works. If so I'll create a PR over in voice-web with the resulting file.

**Comment by MichaelKohler** [comment]
Job running here: https://github.com/Common-Voice/cv-sentence-extractor/runs/743621572?check_suite_focus=true

**Pull request merged**